### PR TITLE
feat(tes): support dual-stack

### PR DESCRIPTION
**Pull request by jcosentino11**
**Issue #, if available:**
resolves https://github.com/aws-greengrass/aws-greengrass-nucleus/issues/1348

**Description of changes:**
Modifies TES web server to listen on `127.0.0.1` and `::1`, allowing it to handle IPv4 and IPv6 requests.

**Why is this change necessary:**
Presently, if a customer component attempts to call TES over IPv6, it will fail because TES is listening on just IPv4 (see issue linked at top).

**How was this change tested:**
- [x] Updated or added new unit tests.
- [ ] Updated or added new integration tests.
- [x] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

Additionally, verified manually in context of containerized lambda function that there was connectivity to TES from ipv4 and ipv6 addresses. 

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [x] I confirm that the change is backwards compatible.
- [ ] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java [hunk 1]
```diff
@@ -31,25 +28,32 @@ public class HttpServerImpl implements Server {
...
         this.executorService = executorService;
     }
 
     @Override
-    public void start() {
-        httpImpl.createContext(URL, credentialRequestHandler);
-        httpImpl.setExecutor(executorService);
-        httpImpl.start();
+    public void start() throws IOException {
+        server = HttpServer.create(new InetSocketAddress(ANY_INTERFACE, configuredPort), 0);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:40]: so this makes us listen on any interface for both ipv4 and ipv6?

Doesn't this also mean that TES is available on the LAN (we don't want that)?

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:40]: Why are we starting a new server within `start()`, that makes start non-idempotent. We could end up leaking server instances it seems.

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:40]: > so this makes us listen on any interface for both ipv4 and ipv6?
> 
> Doesn't this also mean that TES is available on the LAN (we don't want that)?

Yes and yes.  So even though we check Authorization header, reason is we don't want to increase attack surface?  
Will revise

> Why are we starting a new server within `start()`, that makes start non-idempotent. We could end up leaking server instances it seems.

No reason, was just removing network binding from the HttpServerImpl constructor, will revert

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java [hunk 2]
```diff
@@ -26,30 +57,104 @@ public class HttpServerImpl implements Server {
...
+     * @throws BindException if the chosen port is unavailable
+     */
+    private List<HttpServer> createServers(int port) throws IOException {
+        HttpServer serverIPv4 = null;
+        InetSocketAddress addrIPv4 = new InetSocketAddress(ADDR_IPV4, port);
+        try {
+            serverIPv4 = HttpServer.create(addrIPv4, 0);
+        } catch (SocketException e) {
+            if (canBindToAddress(ADDR_IPV4)) {
+                throw e;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:88]: what happens if we throw here? Will it retry? How do we recover?

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java [hunk 3]
```diff
@@ -26,30 +57,104 @@ public class HttpServerImpl implements Server {
...
+            }
+            logger.atDebug().cause(e).kv("address", addrIPv4)
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:90]: this sounds more like a warning, or info

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java [hunk 4]
```diff
@@ -26,30 +57,104 @@ public class HttpServerImpl implements Server {
...
+                new InetSocketAddress(ADDR_IPV6,
+                // use the same port from IPv4 server
+                serverIPv4 == null ? port : serverIPv4.getAddress().getPort());
+        try {
+            serverIPv6 = HttpServer.create(addrIPv6, 0);
+        } catch (SocketException e) {
+            if (canBindToAddress(ADDR_IPV6)) {
+                if (serverIPv4 != null) { // cleanup
+                    stop(serverIPv4);
+                }
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:106]: can we simplify this cleanup logic by just calling `stop()`? and letting stop figure out which server(s) need to stop

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java (hunk 3 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:90]: will bump to info

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java (hunk 2 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:88]: The thought was to let this bubble up as component ERROR and have the component retry.  In the case where port is 0, it's a good chance that the second time would pass.  (Side question--is there ever a time configured port is not the default of 0?  I know we have a `configuration.port` option defined in TokenExchangeService, but it doesn't appear to be set anywhere?)

If error persists and we get into a BROKEN state, then I imagine recovery would be a redeployment, the customer would need to clear up ports on their device

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java (hunk 4 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:106]: yep will rework

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java (hunk 2 above)
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:88]: The port may be set by the customer.

I think we'd want to retry indefinitely if the port is 0 because we should always be able to find some port to use.

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:88]: Sorry the above applies to the similar throw below for ipv6.  When this line throws it should only be possible if configured port != 0 and is already bound, so something is wrong with the configuration or the port needs to be unblocked (so i guess same question still stands if that port configuration is ever expected to change)

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:88]: "should only" since there's a bit of a race---in theory `HttpServer.create(addrIPv4, 0);` could throw when port == 0 when all ports are taken, then `canBindToAddress(ADDR_IPV4)` might return true if a port was freed between those two calls 🤷 

regardless, will add in explicit retrying when port is system-picked and bubble up exception when port is user configured 👍

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java [hunk 5]
```diff
@@ -5,51 +5,202 @@
...
+            RetryUtils.runWithRetry(retryConfig,
+                    () -> {
+                        doCreateServers(port);
+                        return null;
+                    },
+                    "create-http-servers",
+                    logger);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } catch (IOException e) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:109]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java [hunk 6]
```diff
@@ -5,51 +5,202 @@
...
+
+    HttpServerImpl(int port, HttpHandler credentialRequestHandler,
+                   ExecutorService executorService, HttpServerProvider httpServerProvider, SocketFactory socketFactory)
+            throws IOException {
         this.credentialRequestHandler = credentialRequestHandler;
         this.executorService = executorService;
+        this.configuredPort = port;
+        this.httpServerProvider = httpServerProvider;
+        this.socketFactory = socketFactory;
+        createServers(configuredPort);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:75]: having this is the constructor scares me. Creating and starting the server should be moved into `start`.

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java [hunk 7]
```diff
@@ -5,51 +5,202 @@
...
+        try {
+            RetryUtils.runWithRetry(retryConfig,
+                    () -> {
+                        doCreateServers(port);
+                        return null;
+                    },
+                    "create-http-servers",
+                    logger);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:108]: let this method throw the interrupt up.

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java [hunk 8]
```diff
@@ -5,51 +5,202 @@
...
+                .collect(Collectors.toList());
     }
 
     int getServerPort() {
-        return httpImpl.getAddress().getPort();
+        return getServers().stream()
+                .findFirst()
+                .map(HttpServer::getAddress)
+                .map(InetSocketAddress::getPort)
+                .orElseThrow(() -> new IllegalStateException("No servers created"));
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:198]: illegal state is a bad exception. Use checked exceptions or return a reasonable default value.

File: src/test/java/com/aws/greengrass/tes/HttpServerImplTest.java [hunk 9]
```diff
@@ -6,90 +6,340 @@
...
 class HttpServerImplTest {
-    private static final String mockResponse = "Hello World";
 
-    private final ExecutorService executorService = Executors.newFixedThreadPool(1);
+    private static final String MOCK_CREDENTIAL_RESPONSE = "Hello World";
+    private static final String[] hosts = {"localhost", "127.0.0.1", "::1"};
 
-    @Mock
-    private HttpHandler mockHttpHandler;
+    ExecutorService executorService = TestUtils.synchronousExecutorService();
```
Comment (MikeDombo) [inline src/test/java/com/aws/greengrass/tes/HttpServerImplTest.java:59]: shutdown this executor

Comment (jcosentino11) [reply src/test/java/com/aws/greengrass/tes/HttpServerImplTest.java:59]: it's a no-op though? will add for completeness though

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java (hunk 6 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:75]: I did at one point 😅 
will revise and make sure start is idempotent

**Comment by jcosentino11** [comment]
just confirmed passing on UATs

File: src/main/java/com/aws/greengrass/tes/TokenExchangeService.java [hunk 10]
```diff
@@ -85,14 +85,14 @@ public void postInject() {
...
-    protected void startup() {
+    protected void startup() throws InterruptedException {
         logger.atInfo().addKeyValue(PORT_TOPIC, port).addKeyValue(IOT_ROLE_ALIAS_TOPIC, iotRoleAlias)
                 .log("Attempting to start server at configured port {}", port);
         try {
             validateConfig();
-            server = new HttpServerImpl(port, credentialRequestHandler, executor);
-            server.start();
-            logger.atInfo().log("Started server at port {}", server.getServerPort());
+            server.start(port, credentialRequestHandler);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/tes/TokenExchangeService.java:93]: Why did you change the signature of `start()`? Keep it just `start()`. Everything which was set in the constructor is fine, leave that alone. Just don't create and start the server in the constructor.

**Review by MikeDombo — APPROVED** [review]

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java [hunk 11]
```diff
@@ -5,51 +5,203 @@
...
+    }
+
+    @SuppressWarnings({"PMD.AvoidCatchingGenericException", "PMD.AvoidRethrowingException"})
+    private void createServersWithRetry(int port) throws IOException, InterruptedException {
+        RetryUtils.RetryConfig retryConfig = RetryUtils.RetryConfig.builder()
+                .maxAttempt(Integer.MAX_VALUE)
+                .initialRetryInterval(Duration.ofSeconds(1L))
+                .maxRetryInterval(Duration.ofMinutes(1L))
+                .retryableExceptions(Collections.singletonList(SocketException.class))
+                .build();
```
Comment (shaguptashaikh) [inline src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:140]: This doesn't need to be here, you can make this a static constant

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java [hunk 12]
```diff
@@ -5,51 +5,203 @@
...
+     * Create ipv4 and ipv6 servers, both using the specified port.
+     *
+     * @param port chosen port for all created servers
+     * @throws IOException if an I/O error occurs during server creation
+     * @throws BindException if the chosen port is unavailable
+     */
+    private void doCreateServers(int port) throws IOException {
+        InetSocketAddress addrIPv4 = new InetSocketAddress(ADDR_IPV4, port);
+        try {
+            serverIPv4 = httpServerProvider.createHttpServer(addrIPv4, 0);
```
Comment (shaguptashaikh) [inline src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:167]: Is it okay to block on this with infinite retries and never even attempt to create the ipv6 server or is the idea that this can only happen when the port is unavailable so the ipv6 server will also fail anyway? Would we instead benefit from configurable retries the default for which could be infinite retries?

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java [hunk 13]
```diff
@@ -5,51 +5,203 @@
...
+        return Stream.of(serverIPv4, serverIPv6)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    private void createServers(int port) throws IOException, InterruptedException {
+        if (!getServers().isEmpty()) {
+            logger.atInfo().kv("port", port).log("Skipping server creation, servers already running");
+            return;
+        }
```
Comment (shaguptashaikh) [inline src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:121]: How does this work when we need to restart if some configuration for TES has changed?

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:121]: new `HttpServerImpl` is created on each `startup`, https://github.com/aws-greengrass/aws-greengrass-nucleus/pull/1351/files#diff-53c817ca085df370edb8025c1ae9981c0e0d62c164ade69a91290289b3e7d573R95, so it'll respond to config changes.  Technically this path won't ever get triggered today, more of a protection against if `HttpServerImpl#start` were called twice with the same instance.

Comment (shaguptashaikh) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:121]: I see, but looks like that happens on service startup for TES, is config expected to change and take effect without service restart? (I know this is not part of this change and how it existed, just curious how this currently works and if have scope for improvement)

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java (hunk 12 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:167]: I think we should avoid having TES in a false half-state, where v4/v6 are both supported, but we only create one because the other fails due to port being unavailable.  So if v4 fails when we reasonably believe ipv4 is supported on the customer device, then it wouldn't make sense to try v6.  And inversely where v4 succeeds but v6 fails, we'd want to retry until we get both.

Configurable retries could be beneficial but I'm not sure if we'd be retrying much anyway.  I imagine the chance of a system picking a port that gets swiped before the v6 server is created is pretty low.  Maybe it would be better to fail more quickly to signal to the customer that there's some larger issue going on? (cc: @MikeDombo)

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java (hunk 13 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:128]: Ahh, yeah I was just focused on keeping the behavior similar to before, port change taking effect after restart.  I don't see why we couldn't react to port changes in the future.  In the original PR for TES there's a [TODO](https://github.com/aws-greengrass/aws-greengrass-nucleus/pull/240/files#r424732402) for it. Although there's a comment in the code rn saying that [port change shouldn't be allowed](https://github.com/aws-greengrass/aws-greengrass-nucleus/commit/fa3cf396d0001602e291070c970f568f4534eec4#diff-c2b86c6fe185344327806e5cb614804c7e81c6abd7ee63b41f42384fb28d4238R58), curious what that's about

File: src/main/java/com/aws/greengrass/tes/HttpServerImpl.java (hunk 12 above)
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/tes/HttpServerImpl.java:167]: Would we be able to do a lookup for `localhost` and get the address(es) for it? Instead of always trying to bind to ::1, we should only bind to ::1 and 127.0.01 if both are resolved for localhost address. This way, if a system only has ipv4 we should work fine, and the same if the system only has ipv6.
